localgov_eu_cookie_compliance.module

```diff
@@ -217,3 +248,146 @@ function _localgov_eu_cookie_compliance_defer_tracker_scripts(array &$attachment
+ *   Field template variables (s. cookies_video_preprocess_field()).
+ */
+function localgov_eu_cookie_compliance_media_oembed_handler(array &$variables) {
+  $formatter = $variables["element"]["#formatter"];
+  /** @var \Drupal\Core\Field\FieldItemList $field_item */
+  $field_item = $variables['element']['#items'];
+  /** @var \Drupal\media\Entity\Media $media */
+  $media = $variables['element']['#object'];
+  $media_id = $media->id();
+  $variables['attributes']['class'][] = "media-oembed-$media_id";
+  $field_value = $field_item->getValue();
+  if (isset($field_value) && !empty($field_value)) {
+    $field_value = reset($field_value);
+    $video_src = $field_value['value'];
+  }
+
+  // Get the oembed settings lines and split them out.
+  $oembed_settings = Drupal::config('localgov_eu_cookie_compliance.settings')->get('disabled_oembeds') ?? "";
+  $disabled_oembeds = explode("\n", trim($oembed_settings));
+
+  // Convert the domains with optional category into an associative
+  // array of domain to category.
+  $blocked_domains = [];
+  foreach ($disabled_oembeds as $embed) {
+    // Extract the category condition, if present.
+    $parts = explode(':', $embed);
+    if (count($parts) == 1) {
+      $condition = "";
+      $embed = $parts[0];
+    }
+    else {
+      $condition = $parts[0];
+      $embed = end($parts);
+    }
+    $blocked_domains[$embed] = $condition;
+  }
+
+  foreach ($variables["items"] as &$item) {
+    switch ($formatter) {
+      case 'oembed':
+        // Get the video src being embedded.
+        $item["content"]["#attributes"]['data-src'] = $video_src;
+        $parsed = parse_url($video_src);
+
+        // Check against each rule.
+        foreach ($blocked_domains as $domain => $condition) {
+          // Convert the domain specification to a regex by:
+          // 1. Trim and quote the entered domain.
+          // 2. Replace * and ? with .* and . respectively.
+          // 3. Wrap in a case-insensitive search.
+          $regex = str_replace(
+            "\\?",
+            ".",
+            str_replace(
+              "\\*",
+              ".*",
+              preg_quote(trim($domain))
+            )
+          );
+          $regex = "/" . $regex . "/i";
+
+          // If the built regex matches, replace this item.
+          if (preg_match($regex, $parsed["host"])) {
+            // Add a class to the generated video markup.
+            $video = localgov_eu_cookie_compliance_preprocess_field_item_oembed($item);
+            // Render a link instead.
+            localgov_eu_cookie_compliance_preprocess_field_item_link($item);
+            // Record the video markup in the page JSON along with the condition.
+            $elements = [
+              $media->id() => [
+                'video' => Drupal::service('renderer')->render($video, FALSE),
+                'category' => $condition,
+              ],
+            ];
+            // Do not process more rules if one matched.
```

Comment:
There has to be a better way.  In any case, please move this chunk of code to its own function.

Reply:
I agree, especially on moving it to its own function. I very nearly did that already, should have followed that instinct.

When it comes to the logic, let me walk you through the thought process, see if you think we can simplify.

On Essex County Council we didn't have a condition here, it just replaced all embeds and checked them against the marketing cookies category. When it came time to contribute it back, I felt like people would want more control over this. Specifically, I thought it was plausible they'd want to allow some providers (like youtube-nocookies) and not others, or have them grouped differently. I'm not sure that's a serious requirement, if not we could just allow selecting the category for all embeds.

Then, in the first implementation I built, I just did simple base domain matching, but that had the issue of the fall-through case, I wanted to be able to choose whitelist or blacklist, and also a risk of things getting missed due to www. being present or not.

Then, I built this version. I had hoped that there'd be an easy function to do basic pattern matching, I considered simple contains matches, but my head was stuck on https://docs.python.org/3/library/fnmatch.html#fnmatch.fnmatch as that'd what I'd be using if it was Python code. I thought that it would be unlikely that sites would have more than a couple of filters here, most pages would have 0 or 1 video embeds, and they get cached anyway, so I was able to justify to myself making it simpler for users at the expense of runtime complexity.

I'd really value hearing your opinion on what level of customisation we really need.